# Pre-lecture HW - Cayden Wong

### 1. Pick one of the datasets from the ChatBot session(s) of the TUT demo (or from your own ChatBot session if you wish) and use the code produced through the ChatBot interactions to import the data and confirm that the dataset has missing values

Session Link for Q1: https://chatgpt.com/share/1306abdd-53ed-48bb-a048-c6e2bd0829b4

In [17]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

# Count missing values
total_missing = df.isnull().sum().sum()
missing_per_column = df.isnull().sum()
missing_info = pd.DataFrame({
    'Missing Values': missing_per_column,
    'Percentage': df.isnull().mean() * 100
})

print(f"Total missing values in the dataset: {total_missing}")
print("\nMissing values per column:")
print(missing_per_column)
print("\nMissing values information:")
print(missing_info)


Total missing values in the dataset: 12

Missing values per column:
row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

Missing values information:
             Missing Values  Percentage
row_n                     0    0.000000
id                        1    0.255754
name                      0    0.000000
gender                    0    0.000000
species                   0    0.000000
birthday                  0    0.000000
personality               0    0.000000
song                     11    2.813299
phrase                    0    0.000000
full_id                   0    0.000000
url                       0    0.000000


##### Interaction Summary for Q1
Here's a summary of our interaction:

Initial Request:

You asked for a way to count missing values in a dataset from a provided URL using pandas.
Solution Provided:

I provided a step-by-step guide on how to load the dataset and count the missing values.
The steps included loading the dataset, calculating total missing values, missing values per column, and a detailed view with percentages.
Condensation Request:

You asked for the code to be condensed into a single block.
Condensed Code:

I provided a single block of code that accomplishes the tasks of loading the dataset, counting missing values, and displaying the results.
If you need further assistance or have any more questions, feel free to ask!

### 2. Start a new ChatBot session with an initial prompt introducing the dataset you're using and request help to determine how many columns and rows of data a pandas DataFrame has, and then
- use code provided in your ChatBot session to print out the number of rows and columns of the dataset; and,
- write your own general definitions of the meaning of "observations" and "variables" based on asking the ChatBot to explain these terms in the context of your dataset

Session Link for Q2 + Q3: https://chatgpt.com/share/b1d6ee3a-15ce-4074-9282-b548bc7cf547

In [3]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

# Get the number of rows and columns
rows, columns = df.shape

print(f"Number of rows: {rows}")
print(f"Number of columns: {columns}")

Number of rows: 391
Number of columns: 11


**Observations**:
A row within a data set or a single point of information within a collection of data points. Each observation usually carries information about different variables / properties related to the observation.
Examples include: Information about people, objects, events, and places

**Variables**:
A property / charactristic that each observation may or may not have.
Examples include: someones age / height / weight / demographic, time / place of an event, population / GDP / inflation of a country.

### 3) Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to provide these summaries for your dataset

Session Link for Q2 + Q3: https://chatgpt.com/share/b1d6ee3a-15ce-4074-9282-b548bc7cf547

In [5]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

# Print column names
print("=== Column Names ===")
for i, col in enumerate(df.columns):
    print(f"{i + 1}. {col}")

# Print dataset info
print("\n=== Dataset Info ===")
df_info = df.info()

# Print numerical summary
print("\n=== Summary Statistics (Numerical Columns) ===")
numerical_summary = df.describe()
print(numerical_summary)

# Print categorical summary
print("\n=== Summary Statistics (Categorical Columns) ===")
categorical_summary = df.describe(include='object')
print(categorical_summary)

# Show first few rows for preview
print("\n=== First 5 Rows of Data ===")
data_preview = df.head()
print(data_preview)


=== Column Names ===
1. row_n
2. id
3. name
4. gender
5. species
6. birthday
7. personality
8. song
9. phrase
10. full_id
11. url

=== Dataset Info ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB

=== Summary Statistics (Numerical Columns) ===
            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50

##### Interaction Summary for Q2 + Q3
Sure! Here’s a summary of our interaction:

1. **Dataset Overview**:
   - You are working with a dataset available at [this URL](https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv).
   - You wanted to know how to determine the number of rows and columns in the dataset using pandas.

2. **Determining Rows and Columns**:
   - You can use the `shape` attribute of a pandas DataFrame to get the number of rows and columns.
   - Example code:
     ```python
     import pandas as pd

     # Load the dataset
     url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
     df = pd.read_csv(url)

     # Get the number of rows and columns
     rows, columns = df.shape

     print(f"Number of rows: {rows}")
     print(f"Number of columns: {columns}")
     ```

3. **Understanding Observations and Variables**:
   - **Observations**: Individual data points or rows in the dataset (e.g., each villager).
   - **Variables**: Attributes or columns in the dataset that describe observations (e.g., name, age, species).

4. **Extracting Information About Columns**:
   - To get a summary of columns, you can use methods like:
     - `df.info()` for a concise summary.
     - `df.describe()` for numerical columns' statistics.
     - `df.describe(include='object')` for categorical columns' statistics.
     - `df.columns` to list column names.
     - `df.head()` to preview the first few rows.

5. **Formatting Output**:
   - You can format the output to be more readable by clearly labeling sections and providing structured information.
   - Example of structured output:
     ```python
     print("=== Column Names ===")
     for i, col in enumerate(df.columns):
         print(f"{i + 1}. {col}")

     print("\n=== Dataset Info ===")
     df_info = df.info()

     print("\n=== Summary Statistics (Numerical Columns) ===")
     numerical_summary = df.describe()
     print(numerical_summary)

     print("\n=== Summary Statistics (Categorical Columns) ===")
     categorical_summary = df.describe(include='object')
     print(categorical_summary)

     print("\n=== First 5 Rows of Data ===")
     data_preview = df.head()
     print(data_preview)
     ```

6. **Understanding Numerical Summary**:
   - The numerical summary provides key statistics for numerical columns:
     - **Count**: Number of non-null values.
     - **Mean**: Average value.
     - **Standard Deviation (std)**: Measure of spread.
     - **Min**: Minimum value.
     - **25%**, **50%** (median), **75%**: Quartiles.
     - **Max**: Maximum value.
   - These statistics help assess data completeness, distribution, and variability.

Feel free to ask if you need any further clarification or additional details!

### 4) If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, explain (perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by df.shape and what is reported by df.describe() with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column

Session Link for Q4: https://chatgpt.com/share/fd9be579-b2d3-4bc0-9e31-3d70e9666d16

In [16]:
import pandas as pd

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

print(df.shape)

print(df.describe(include='all'))
# print(df.describe())

(891, 15)
          survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class  who adult_male deck  embark_town alive  \

In [14]:
df.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [13]:
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

**Why is the number of variables different that what is described in `.shape` using `.describe()`?**
When using `.describe()` without any additional parameters, the function only returns data about variables that contain only integer values. This means columns that include information such as name, class, and embark_town are not included when using `.describe()` leading to a lower percived number of variables. This can be fixed to include all types by adding the following additional argument: `.describe(include='all')`.

**Why is the count number different between variables?**
The count number is different to the number described by `.shape` as it does not count missing values contained within the dataset. For example, the variable survived returns  a count of `891` (the same as described by `.shape`) while age only returns a count of `714`. This likely indicates that there is missing data about the age of some of the observations within the dataset, most likely information that was unobtainable because the passenger may have not survived.

### 5) Use your ChatBot session to help understand the difference between the following and then provide your own paraphrasing summarization of that difference

Session Link for Q4 + Q5: https://chatgpt.com/share/fd9be579-b2d3-4bc0-9e31-3d70e9666d16

**Attribute** (Example: `df.shape`)
An attribute is an intrinsic property of some object that stores information or data about that object (in the case of `.shape`, it is a property describing the dimensions of a pandas DataFrame object). It is not a function call and does not need to be called with any parameters.

**Method** (Example: `df.describe()`)
A method is an property of some object that performs and operation / returns information about said object (for .describe(), this is a method of the pandas DataFrame object). Since it is a function call, it can take in additional paramaters within the `()`.

##### Interaction summary for Q4 + Q5
In this conversation:

You were working with a dataset from the Titanic (https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv) using pandas and wanted to know how to make df.describe() output summary statistics for all data types, not just numerical ones. I suggested using the include='all' parameter in df.describe() to include all data types (numerical, categorical, etc.).

You asked why the top row in df.describe(include='all') sometimes reports NaN as its value. I explained that this can happen when:

The column contains mostly or entirely NaN values.
There are mixed data types in the column.
NaN is the most frequent value in that column.
You asked about the difference between an attribute and a method. I explained that an attribute holds data or properties of an object (like .shape), while a method is a function that performs an action (like .describe()), requiring parentheses to invoke it.

Let me know if you'd like more details or clarification on any part!